File: javascript/lesson 15/longestIncreasingSubsequence.js
// https://leetcode.com/problems/longest-increasing-subsequence/

function lengthOfLIS(nums) {
    const piles = [];
       
   for (let num of nums) {
       const toInsert = searchInsert(piles, num);
       if (toInsert == piles.length) {
           piles.push(num);
       } else {
           piles[toInsert] = num;
       }
   }
       
   return piles.length;
}
   
function searchInsert(piles, target) {
   let lo = 0;
   let hi = piles.length - 1;

   while (lo <= hi) {
       const mid = lo + ((hi - lo) >> 1);
       const value = piles[mid];
       if (value == target) {
           return mid;
       } else if (value < target) {
           lo = mid + 1;
       } else {
           hi = mid - 1;
       }
   }

   return lo;
}

File: ruby/lesson 15/longest_increasing_subsequence.rb
# https://leetcode.com/problems/longest-increasing-subsequence/

def length_of_lis(nums)
    piles = []    
    nums.each do |num|
        to_insert = search_insert(piles, num)
        if to_insert == piles.length
            piles.push(num)
        else
            piles[to_insert] = num
        end
    end    
    piles.length
end

def search_insert(nums, target)
    lo = 0
    hi = nums.length - 1

    while lo <= hi
        mid = lo + (hi - lo) / 2
        next hi = mid - 1 if nums[mid] > target
        next lo = mid + 1 if nums[mid] < target
        return mid
    end

    lo
end